File: bench/langs/ledger.go
package ledger

// Limit caps a single charge.
const Limit = 5000

// Charge bills an account and returns what is left.
func Charge(account string, amount int) int {
	balance := lookup(account)
	if balance < amount {
		return refuse(account)
	}
	return commit(account, balance-amount)
}

func lookup(account string) int {
	return read(account)
}

func refuse(account string) int {
	warn(account)
	return 0
}

func commit(account string, left int) int {
	write(account, left)
	return left
}

File: bench/langs/ledger.rs
//! Keeps the running balance for one account.

/// Caps a single charge.
pub const LIMIT: i64 = 5000;

pub struct Ledger {
    owner: String,
    entries: Vec<i64>,
}

impl Ledger {
    /// Bills the account and returns what is left.
    pub fn charge(&mut self, amount: i64) -> Option<i64> {
        if amount > LIMIT {
            return self.refuse(amount);
        }
        self.entries.push(amount);
        Some(self.total())
    }

    fn total(&self) -> i64 {
        sum_of(&self.entries)
    }

    fn refuse(&mut self, amount: i64) -> Option<i64> {
        warn(&self.owner);
        None
    }
}
